File: sudo/src/main/java/com/roman/gurdan/sudo/game/action/MirrorManager.java
package com.roman.gurdan.sudo.game.action;

import java.util.Stack;

public class MirrorManager implements IMirrorManager {

    private static final Object _object = new Object();
    private Stack<AMirror> stack;

    public MirrorManager() {
        stack = new Stack<>();
    }

    @Override
    public void addMirror(AMirror action) {
        synchronized (_object) {
            if (action != null && action.valid())
                stack.push(action);
        }
    }

    @Override
    public boolean canUndo() {
        synchronized (_object) {
            return stack.size() > 0;
        }
    }

    @Override
    public AMirror undo() {
        synchronized (_object) {
            return stack.pop();
        }
    }

    @Override
    public void clear() {
        synchronized (_object) {
            stack.clear();
        }
    }

}

File: base/src/main/java/com/roman/garden/sudo/base/action/MirrorManager.kt
package com.roman.garden.sudo.base.action

import java.util.*

class MirrorManager : IMirrorManager {

    override var stack: Stack<IMirror> = Stack()
//    val mutex = Mutex()

    override fun addMirror(iMirror: IMirror) {
//        mutex.lock {
        if (iMirror.valid()) stack.push(iMirror)
//        }
    }

    override fun hasUndo(): Boolean = stack.size > 0

    override fun undo(): IMirror? = stack.pop()

    override fun clear() = stack.clear()
}